llmfoundry/data/finetuning/dataloader.py

```diff
@@ -199,28 +159,92 @@ def build_finetuning_dataloader(cfg: DictConfig,
                         +
                         f'of samples in your dataset to at least {minimum_dataset_size}.'
                     )
-
-        assert dataset is not None
-        dl = DataLoader(
-            dataset,
-            collate_fn=collate_fn,
-            batch_size=dataloader_batch_size,
+        sampler = dist.get_sampler(dataset,
             drop_last=cfg.drop_last,
-            sampler=dist.get_sampler(dataset,
-                                     drop_last=cfg.drop_last,
-                                     shuffle=cfg.dataset.shuffle),
-            num_workers=cfg.num_workers,
-            pin_memory=cfg.get('pin_memory', True),
-            prefetch_factor=cfg.get('prefetch_factor', 2),
-            persistent_workers=cfg.get('persistent_workers', True),
-            timeout=cfg.get('timeout', 0),
-        )
+            shuffle=cfg.dataset.shuffle)
+    else:
+        # Build streaming dataset
+        dataset = _build_streaming_dataset(cfg.dataset, tokenizer, device_batch_size)
+    
+    assert dataset is not None
+    dl = DataLoader(
+        dataset,
+        collate_fn=collate_fn,
+        batch_size=dataloader_batch_size,
+        drop_last=cfg.drop_last,
+        sampler=sampler,
+        num_workers=cfg.num_workers,
+        pin_memory=cfg.get('pin_memory', True),
+        prefetch_factor=cfg.get('prefetch_factor', 2),
+        persistent_workers=cfg.get('persistent_workers', True),
+        timeout=cfg.get('timeout', 0),
+    )
 
     token_counting_func = get_tokens_per_batch_func(
         pad_token_id=tokenizer.pad_token_id)
 
     return DataSpec(dataloader=dl, get_num_tokens_in_batch=token_counting_func)
 
+def _build_preprocessing_function(dataset_name_or_path: str, proto_preprocessing_fn: Optional[Union[dict, DictConfig, str]]):
```

Comment:
nit: method comments are missing. Please also check other methods.

Reply:
Will do! still wip, will make sure we have some good code documentation once it's ready for review!